libyul/optimiser/SSATransform.cpp

```diff
@@ -165,24 +165,24 @@ class IntroduceControlFlowSSA: public ASTModifier
 	NameDispenser& m_nameDispenser;
 	std::set<YulString> const& m_variablesToReplace;
 	/// Variables (that are to be replaced) currently in scope.
-	std::set<YulString> m_variablesInScope;
-	/// Set of variables that do not have a specific value.
-	std::set<YulString> m_variablesToReassign;
+	std::vector<YulString> m_variablesInScope;
+	/// Variables that do not have a specific value.
+	std::vector<YulString> m_variablesToReassign;
```

Comment:
Maybe a `set` with a custom comparator that compares based on `YulString::str()` would be a cleaner solution here? This would give us a deterministic iteration order while still being a set with logarithmic complexity and not requiring operation through special helpers (which someone could forget to user when modifying the code later, since normal non-unique insertion is still available).

Reply:
Yes, you're right. I thought `YulString` did not have a deterministic sort order but it does so we can't do it the way I suggested above.

Still, having to remember to use these helpers to manipulate the vector like that just seems too error-prone to me. It would be fine it this was a separate data structure that hides the original `push_back()`&co but it's not.

How about a different approach - instead of keeping duplicates out on every insertion, how about we keep them in the vector but skip them later during iteration? We could have a nice and reusable `unique()` function that would give you a vector with only unique items (I actually wanted to suggest `ranges::unique()` originally, but I see it removes only adjacent duplicates).

Reply:
The reason I added the helper functions was to reduce code duplication, and then to eventually re-use them in other fixes of this type (retroactively, since we've already had at least one PR that used this same approach). I would prefer to merge this as is, and then we can decide on one of the Wednesday calls what would be the ideal way to address this in the future, and then we can re-factor in a separate PR.

Reply:
The problem I have with it is that "a vector for which you have to remember to keep items unique by hand" does not seems like a great data structure to have and I'd rather avoid generalizing that pattern without also making it safer. It's dangerous and if we're not careful we'll just turn one kind of easy to make, hard to find bug into another.

If you really want to keep this pattern then at least wrap it in some kind of custom container. I.e. define a `UniqueVector` class where these helpers are methods and it only provides access to the `const` version of the underlying vector. Should be quite short and without too much boilerplate if you simply put it in `CommonData.h` where the helpers currently are. This way we'll be able to keep it somewhat contained.

We might even, in the future, store an extra `set` inside to make lookups more efficient. The linear lookups in the current version bother me a bit, but I'm ignoring that because the set of variables in SSA will probably never be that big and managing the set by hand would make managing it even more error-prone. But if you promote it to more general pattern that we'll use elsewhere that becomes a consideration again.

Reply:
Again, this is the same approach used in https://github.com/ethereum/solidity/pull/14311, which wasn't *ideal* then either; I'd rather not waste time implementing something without getting to a consensus as to what the best approach for this would be (or at least, and combination of best and feasible). @ekpyron?

Reply:
The important difference is that your PR generalizes that approach, which is not a good idea, because it will just make it easier for this unsafe pattern to spread.

I'm not demanding a perfect solution here, I'm just saying that there are ways to make it safer and some of them require very little extra effort. Even something as simple as this would do:

```c++
template <typename T>
class UniqueVector
{
public:
	std::vector<T> content() const { return m_content; }
	bool contains(T const& _value) const { return m_content.count(_value) != 0; }

	void append(T _value)
	{
		if (!contains(m_content, _value))
			m_content.emplace_back(std::move(_value));
	}

	void appendMissing(std::vector<T> const& _values)
	{
		for (auto&& value: _values)
			append(m_content, value);
	}

	void removeAll(std::vector<T> const& _values)
	{
		for (auto const& value: _values)
			m_content.erase(std::find(m_content.begin(), m_content.end(), _value));
	}

private:
	std::vector<T> m_content;
};
```

which is basically your solution but in a safety wrapper.